src/gluonts/torch/model/wavenet/module.py

```diff
@@ -0,0 +1,569 @@
+        )
+        return full_features
+
+    def target_feature_embedding(
+        self, target: torch.Tensor, features: torch.Tensor
+    ) -> torch.Tensor:
+        """Provides a joint embedding for the target and features.
+
+        Parameters
+        ----------
+        target
+            Full target of shape (batch_size, sequence_length)
+        features
+            Full features of shape (batch_size, num_features, sequence_length)
+
+        Returns
+        -------
+            A tensor containing a joint embedding of target and features.
+            Shape: (batch_size, n_residue, sequence_length)
+        """
+        out = self.target_embedder(target)
+        out = torch.transpose(out, 1, 2)
+        out = torch.cat([out, features], dim=1)
+        out = self.conv_project(out)
+        return out
+
+    def base_net(
+        self,
+        inputs: torch.Tensor,
+        prediction_mode: bool = False,
+        queues: List[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, List[torch.Tensor]]:
+        """Forward pass through the WaveNet.
+
+        Parameters
+        ----------
+        inputs
+            A tensor of inputs
+            Shape: (batch_size, n_residual_channels, sequence_length)
+        prediction_mode, optional
+            Flag indicating whether the network is being used
+            for prediction, by default False
+        queues, optional
+            Convolutional queues containing past computations.
+            This speeds up predictions and must be provided
+            during prediction mode. See [Paine et al., 2016] for details,
+            by default None
+
+        [Paine et al., 2016] "Fast wavenet generation algorithm."
+           arXiv preprint arXiv:1611.09482 (2016).
+        Returns
+        -------
+            A tensor containing the unnormalized outputs of the network of
+            shape (batch_size, pred_length, num_bins) and a list containing the
+            convolutional queues for each layer. The queue corresponding to
+            layer `l` has shape: (batch_size, n_residual_channels, 2^l).
+        """
+        if prediction_mode:
+            assert (
+                queues is not None
+            ), "Queues cannot be empty in prediction mode!"
+
+        skip_outs = []
+        queues_next = []
+        out = inputs
+        for i, layer in enumerate(self.residuals):
+            skip, out = layer(out)
+            if prediction_mode:
+                trimmed_skip = skip
+                if i + 1 < len(self.residuals):
+                    out = torch.cat([queues[i], out], dim=-1)
+                    queues_next.append(out[..., 1:])
+            else:
+                trimmed_skip = skip[..., self.trim_lengths[i] :]
+            skip_outs.append(trimmed_skip)
+
+        y = torch.stack(skip_outs).sum(dim=0)
+        y = self.output_act(y)
+        y = self.conv1(y)
+        y = self.output_act(y)
+        y = self.conv2(y)
+        logits = y.transpose(1, 2)
```

Comment:
@abdulfatir I think the problem in your test appears because of this transpose here: if you compute the grad of `y` w.r.t. `inputs`, you will get correct causal masking.

Reply:
Ah, right! I need to sum over `-1`, not `-2`.

Reply:
I really like this test @abdulfatir (and good catch @shchur). I remember we had off-by-one errors a few times between training and prediction target. Maybe one can make the test re-usable for other models and also check that no time point is skipped.

Reply:
That's a good idea. Let me check how we can add these tests to other models.

Reply:
> I remember we had off-by-one errors a few times

Two hard things in computer science: cache invalidation, naming things, and off-by-one errors